doc/ext-pake/proposed/spake2plus-proposal-draft.md

```diff
@@ -0,0 +1,66 @@
+Introduction
+------------
+
+This document proposes protocol flow and high-level API changes required in PSA Cryptography API 1.1 PAKE Extension, to support the SPAKE2+ protocol.
+
+SPAKE2+ Version
+---------------
+
+SPAKE2+, an Augmented PAKE Draft 02, 10 December 2020 is considered for proposal.
+
+Link : [https://tools.ietf.org/pdf/draft-bar-cfrg-spake2plus-02.pdf](https://tools.ietf.org/pdf/draft-bar-cfrg-spake2plus-02.pdf)
+
+Remarks
+
+*   SPAKE2+, an Augmented PAKE Draft 08, 5 May 2022 is the latest draft version. Link : [https://datatracker.ietf.org/doc/pdf/draft-bar-cfrg-spake2plus-08](https://datatracker.ietf.org/doc/pdf/draft-bar-cfrg-spake2plus-08)
+*   Shared Secret Key generation is not compatible between Draft 02 and 08.
+*   As most SPAKE2+ implementations e.g. Matter Specification Version 1.0 are based on Draft 02, this version is being considered for better interoperability.
+
+Expected PAKE API Flow :
+------------------------------------
+![](SPAKE2PLUS.svg)
+
+
+
+#### Out of band set up
+
+SPAKE2+ has an offline initialization step where both parties may share parties identities and additional data (the context). This step is considered out of scope but these shared information may be required for SPAKE2+ protocol execution.
+
+**Shared Information** : ProverID, VerifierID and Context.
+
+**Input methods** :
+
+*   Existing API's (when user is Prover)
+    *   `psa_pake_set_user()` to input ProverID
+    *   `psa_pake_set_peer()` to input VerifierID
+*   Proposed new API
+    *   `psa_pake_set_context()` to input context (additional data).
+
+#### Password, PBKDF Output and Verification value (W0, W1, L)
+
+SPAKE2+ requires PBKDF Output and Verification value (W0, W1, L) generated from password as below :
+
+ `w0s || w1s = PBKDF(len(pw) || pw || len(A) || A || len(B) || B)`  
+ `W0 = w0s mod p`  
+ `W1 = w1s mod p`  
+ `L = w1*P`
+
+Computing these values is considered Out of Scope for PAKE API as this a registration phase which could happen out of actual protocol flow.
```

Comment:
The PBKDF part is covered well by the key derivation API. I agree that we should provide a tool for computing `w0`, `w1` and `L`.

There are several ways to go about this, here is a couple of options that come to mind:
- Provide a new function in the PAKE API to generate them (might be hard to make it universal enough to support other PAKEs as well)
- Add input and output steps to handle the generation (the input/output flow is confusing enough already as it is, this would make it even worse)
- Add a new algorithm to the key derivation API (each PAKE that needs this would take up a point in the key derivation algorithm identifier space and separating the 3 outputs would be error prone)

Reply:
Summarizing the discussion from above and other comments, I see two options.
1. Accept only `w0 || w1` on the prover side and `w0 || L` on verifier side
    - Pros: Can use `psa_pake_set_password_key()` directly.
    - Cons: Bignum calculation of obtaining `w0`, `w1`, and `L` needs to be done by user.
2. Accept `w0s || w1s`. Can also accept instead `w0`, `w1` and `L`.
   - Pros: Addresses concerns of bignum calculations. Can directly feed output of PBKDF operation to PAKE API.
   - Cons: Need to create new function or expand existing `psa_pake_set_password_key()` to accept multiple inputs by adding steps/input types. Can create a generic `psa_pake_input_key()` which accepts keys.
 
In both options the input is treated as a secret.
I am personally leaning towards option 2. We can generalize the input function for keys/secrets e.g. `psa_pake_input_key()` and add the input steps accordingly. This option offers the greatest flexibility in usage.